File: DP/uglyNumberII.cpp
/**
0264 Ugly Number II
2 3 5 corresponding pointers p2, p3, p5
dp[i] = min(dp[p2] * 2, dp[p3] * 3, dp[p5] * 5)
Then move the pointer by 1
*/
class Solution 
{
public:
    int nthUglyNumber(int n) 
    {
        vector<int> dp_array(n + 1, 0);
        dp_array[1] = 1;
        int p2 = 1, p3 = 1, p5 = 1;
        for (int i = 2; i < n + 1; i++) 
        {
            int num2 = dp_array[p2] * 2, num3 = dp_array[p3] * 3, num5 = dp_array[p5] * 5;
            dp_array[i] = min(min(num2, num3), num5);
            if (num2 == dp_array[i]) 
            {
                p2++;
            }
            if (num3 == dp_array[i]) 
            {
                p3++;
            }
            if (num5 == dp_array[i]) 
            {
                p5++;
            }
        }
        return dp_array[n];
    }
};

File: DP/uglyNumberII.py
'''
0264 Ugly Number II
DP idea & pointers.
'''
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        DP = [0] * (n + 1)
        DP[1] = 1
        p2 = p3 = p5 = 1
        for i in range(2, n + 1):
            two = DP[p2] * 2
            three = DP[p3] * 3
            five = DP[p5] * 5
            DP[i] = min(two, three, five)
            if (DP[i] == two):
                p2 += 1
            if (DP[i] == three):
                p3 += 1
            if (DP[i] == five):
                p5 += 1
        return DP[n]